### PR TITLE
Met phi, genweight , tau DM

### PR DESCRIPTION
**Pull request by easilar**

### CODE REVIEW

**Review by easilar — COMMENTED** [review]
looks good

File: nanoaodframe/src/NanoAODAnalyzerrdframe.cpp [hunk 1]
```diff
@@ -156,7 +156,7 @@ void NanoAODAnalyzerrdframe::setupAnalysis() {
             WeightCalculatorFromHistogram* _puweightcalc_plus = new WeightCalculatorFromHistogram(_hpumc, _hpudata_plus);
             WeightCalculatorFromHistogram* _puweightcalc_minus = new WeightCalculatorFromHistogram(_hpumc, _hpudata_minus);
 
-            _rlm = _rlm.Redefine("unitGenWeight","genWeight != 0 ? genWeight/abs(genWeight) : 0")
+            _rlm = _rlm.Redefine("unitGenWeight","genWeight != 0 ? genWeight/genEventSumw : 0")
```
Comment (minerva1993) [inline nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:159]: I don't think this is a normalization, genEventSumw is a single number for a nanoaod root file, no?

File: nanoaodframe/src/TopLFVAnalyzer.cpp [hunk 2]
```diff
@@ -25,9 +25,11 @@ void TopLFVAnalyzer::defineCuts() {
     //addCuts("nmuonpass == 1 && nvetoelepass == 0 && nvetomuons == 0","0");
     // Wil remove PV cut after new skim
     addCuts("nmuonpass == 1 && nvetoelepass == 0 && nvetomuons == 0 && PV_npvsGood > 0","0");
+    //addCuts("nmuonpass == 1 && nvetoelepass == 0 && nvetomuons == 0","0");
     addCuts("ncleantaupass == 1", "00");
     addCuts("mutau_charge < 0", "000");
     addCuts("ncleanjetspass >= 3", "0000");
+    //addCuts("mutau_dR < 1.2", "00000");
```
Comment (minerva1993) [inline nanoaodframe/src/TopLFVAnalyzer.cpp:32]: This is better to delete

File: nanoaodframe/src/TopLFVAnalyzer.cpp [hunk 3]
```diff
@@ -80,18 +82,18 @@ void TopLFVAnalyzer::defineMoreVars() {
...
-    addVar({"chi2_wjet1_idx", "top_reco_whad[3]",""});
-    addVar({"chi2_wjet2_idx", "top_reco_whad[4]",""});
-    addVar({"chi2_SMW", "top_reco_whad[5]",""});
-    addVar({"chi2_SMTop", "top_reco_whad[6]",""});
-
-    defineVar("top_reco_prod", ::top_reco_products_STLFV, {"cleanjet4vecs","muon4vecs","cleantau4vecs","top_reco_whad"});
-    addVar({"chi2_wqq_dEta","top_reco_prod[0]",""});
-    addVar({"chi2_wqq_dPhi","top_reco_prod[1]",""});
-    addVar({"chi2_wqq_dR","top_reco_prod[2]",""});
+   addVar({"chi2", "top_reco_whad[0]",""});
```
Comment (minerva1993) [inline nanoaodframe/src/TopLFVAnalyzer.cpp:85]: indentation...

File: nanoaodframe/src/NanoAODAnalyzerrdframe.cpp [hunk 4]
```diff
@@ -508,7 +508,56 @@ void NanoAODAnalyzerrdframe::setupJetMETCorrection(string globaltag, std::vector
                 mety -= (jetptsafter[i] - jetptsbefore[i])*sin(jetphis[i]);
             }
         }
-        return (float(sqrt(metx*metx + mety*mety)));
+
```
Comment (minerva1993) [inline nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:511]: I think we need to perform this for metPhiUnc and metUnc as well...

Comment (easilar) [reply nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:511]: I know but this is the correction, unc will come in the next.

File: nanoaodframe/src/NanoAODAnalyzerrdframe.cpp (hunk 1 above)
Comment (easilar) [reply nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:159]: I also dig this up a bit, there is no clear explanation on if it is per file or not. But there is a big chance that it is. So we need to cross check with full sample values before skimming. 
But for now this more correct than simply dividing with abs value. 
[1] https://github.com/cms-nanoAOD/nanoAOD-tools/issues/120

File: nanoaodframe/src/TopLFVAnalyzer.cpp (hunk 2 above)
Comment (easilar) [reply nanoaodframe/src/TopLFVAnalyzer.cpp]: done

File: nanoaodframe/src/NanoAODAnalyzerrdframe.cpp (hunk 1 above)
Comment (minerva1993) [reply nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:159]: genEventSumw is stored in Runs branch, not in Events. This is one number for a file. Unitgenweight is a base event weight applied for each event.

File: nanoaodframe/src/NanoAODAnalyzerrdframe.cpp (hunk 4 above)
Comment (minerva1993) [reply nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:511]: I understand. But this information is filled in during skimming (L.179) Which means we need to re-skim for this at some moment.

File: nanoaodframe/src/NanoAODAnalyzerrdframe.cpp (hunk 1 above)
Comment (easilar) [reply nanoaodframe/src/NanoAODAnalyzerrdframe.cpp:159]: Yes unitgen weight is per event , and the best way is to calculate sum of all gen weights per dataset and normilise the event weight by that. The event weight in principle should be "genweight/ sum of all genweights before skimmming ".

**Review by minerva1993 — APPROVED** [review]
good to go

**Pull request merged**